File: leetcode/0077.combinations/typescript/001.backtracking.ts
function combine(n: number, k: number): number[][] {
    const res: number[][] = [];

    const backtrack = (start: number, combination: number[]) => {
        if(combination.length === k) {
            res.push([...combination])
        }

        for(let i = start; i < n; i++) {
            combination.push(i+1);
            backtrack(i+1, combination);
            combination.pop();
        }
    }

    backtrack(0, []);

    return res;
};

File: leetcode/0077.combinations/go/001.backtracking.go
func combine(n int, k int) [][]int {
    res := [][]int{}

    var backtrack func(int, []int)
    backtrack = func(start int, comb []int) {
        if len(comb) == k {
            tmp := make([]int, k)
            copy(tmp, comb)
            res = append(res, tmp)
        }

        for i := start; i < n; i++ {
            comb = append(comb, i+1)
            backtrack(i+1, comb)
            comb = comb[:len(comb)-1]
        }
    }

    backtrack(0, []int{})

    return res
}
